crates/agglayer-errors/src/lib.rs

```diff
@@ -0,0 +1,209 @@
+        'result: {
+            let expr: $crate::eyre::Report = $expr;
+            $(
+                #[allow(unused_variables)] // Pattern variables are unused from this downcast_ref
+                if let Some($pat) = expr.downcast_ref::<$ty>() {
+                    let Ok($pat) = expr.downcast::<$ty>() else {
+                        // This should never happen because we just checked downcast_ref
+                        unreachable!("downcast_ref succeeded but downcast failed");
+                    };
+                    let res = $handler;
+                    break 'result res;
+                }
+            )*
+            $(
+                let $default_pat = expr;
+                $default_handler
+            )?
+        }
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    #[derive(Debug, thiserror::Error)]
+    #[error("Foo")]
+    struct Foo;
+
+    #[derive(Debug, thiserror::Error)]
+    #[error("Bar")]
+    struct Bar(&'static str);
+
+    #[derive(Debug, thiserror::Error)]
+    #[error("Quux")]
+    struct Quux;
+
+    fn use_match_err_to_make_a_value<E: Into<eyre::Report>>(err: E) -> String {
+        let err: eyre::Report = err.into();
+        match_err!(err,
+            Foo: _ => "foo".to_string(),
+            Bar: Bar("foo") => "bar/foo".to_string(),
+            Bar: Bar(s) => format!("bar({s})"),
+            @default: _ => "other".to_string(),
+        )
+    }
+
+    #[test]
+    fn match_err_to_make_a_value() {
+        // Basic tests
+        assert_eq!(use_match_err_to_make_a_value(Foo), "foo");
+        assert_eq!(use_match_err_to_make_a_value(Bar("baz")), "bar(baz)");
+        assert_eq!(use_match_err_to_make_a_value(Bar("foo")), "bar/foo");
+
+        // Default works
+        assert_eq!(use_match_err_to_make_a_value(Quux), "other");
+        assert_eq!(use_match_err_to_make_a_value(eyre::eyre!("baz")), "other");
+
+        // Wrapping with context has no impact
+        assert_eq!(
+            use_match_err_to_make_a_value(eyre::Report::from(Foo).wrap_err("with some context")),
+            "foo"
+        );
+
+        // Wrapping with another match-able error shows the first match among the
+        // match_err branches.
+        // Not necessarily the best choice, but it should be good enough for now.
+        assert_eq!(
+            use_match_err_to_make_a_value(eyre::Report::from(Foo).wrap_err(Bar("baz"))),
+            "foo"
+        );
+        assert_eq!(
+            use_match_err_to_make_a_value(eyre::Report::from(Bar("baz")).wrap_err(Foo)),
+            "foo"
+        );
+
+        // Wrapping an unknown error with a match-able error does show the match-able
+        // error
+        assert_eq!(
+            use_match_err_to_make_a_value(eyre::Report::from(Quux).wrap_err(Foo)),
+            "foo"
+        );
+        assert_eq!(
+            use_match_err_to_make_a_value(eyre::Report::from(Quux).wrap_err(Foo).wrap_err(Quux)),
+            "foo"
+        );
```

Comment:
Again, we should not just partially match the inner error, if we can not match the outer one.